sweep/sweeper.go

```diff
@@ -1878,3 +1884,141 @@ func (s *UtxoSweeper) IsSweeperOutpoint(op wire.OutPoint) bool {
 
 	return found
 }
+
+// markInputSwept marks the given input as swept by the tx. It will also notify
+// all the subscribers of this input.
+func (s *UtxoSweeper) markInputSwept(inp *SweeperInput, tx *wire.MsgTx) {
+	log.Debugf("Marking input as swept: %v from state=%v", inp.OutPoint(),
+		inp.state)
+
+	inp.state = Swept
+
+	// Signal result channels.
+	s.signalResult(inp, Result{
+		Tx: tx,
+	})
+
+	// Remove all other inputs in this exclusive group.
+	if inp.params.ExclusiveGroup != nil {
+		s.removeExclusiveGroup(*inp.params.ExclusiveGroup)
+	}
+}
+
+// handleThirdPartySpent takes an input and its spending tx. If the spending tx
+// cannot be found in the sweeper store, the input will be marked as fatal,
+// otherwise it will be marked as swept.
+func (s *UtxoSweeper) handleThirdPartySpent(inp *SweeperInput, tx *wire.MsgTx) {
+	op := inp.OutPoint()
+	txid := tx.TxHash()
+
+	isOurTx, err := s.cfg.Store.IsOurTx(txid)
+	if err != nil {
+		log.Errorf("Cannot determine if tx %v is ours: %v", txid, err)
+		return
+	}
+
+	// If this is our tx, it means it's a previous sweeping tx that got
+	// confirmed, which could happen when a restart happens during the
+	// sweeping process.
+	if isOurTx {
+		log.Debugf("Found our sweeping tx %v, marking input %v as "+
+			"swept", txid, op)
+
+		// We now use the spending tx to update the state of the inputs.
+		s.markInputSwept(inp, tx)
+
+		return
+	}
+
+	// Since the input is spent by others, we now mark it as fatal and won't
+	// be retried.
+	s.markInputFatal(inp, ErrRemoteSpend)
+
+	log.Debugf("Removing descendant txns invalidated by (txid=%v): %v",
+		tx.TxHash(), lnutils.SpewLogClosure(tx))
+
+	// Construct a map of the inputs this transaction spends.
+	inputsSpent := make(map[wire.OutPoint]struct{}, len(tx.TxIn))
+	for _, txIn := range tx.TxIn {
+		inputsSpent[txIn.PreviousOutPoint] = struct{}{}
+	}
+
+	err = s.removeConflictSweepDescendants(inputsSpent)
```

Comment:
I think at this point, we should try to send some event into the fee bumper and/or reliable broadcaster that they can _give up_ on trying to broadcast the transaction. This should cut down on spammy logs, and also potentially help u avoid some heisenbugs related to this and incorrect utxo states.

Reply:
I think we already `CancelRebroadcast` in `removeConflictSweepDescendants` tho. And by the time we reach it here, it means the either,
- a `TxUnknownSpent` is sent from the fee bumper and received by the sweeper, which is handled in `handleBumpEventTxUnknownSpent`, which means the fee bumper has already removed this record from its map.
- during restart, an already spent input is offered to the sweeper, which will be handled in `checkInputSpent`, and since it's marked as either `Fatal` or `Swept` here, this input won't be aggregated and sent to the fee bumper.

PS, TILed the term "heisenbugs"😆